File: python/dmp_param/DMPUnrollInitParams.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
import copy
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_state/'))
from DMPState import *
from DMPTrajectory import *
from TauSystem import *

class DMPUnrollInitParams:
    'Class for DMP Unroll Initialization Parameters.'
    
    def __init__(self, critical_dmptraj_init, tau_init=None, robot_task_servo_rate=None, is_zeroing_out_velocity_and_acceleration=True):
        traj_size = critical_dmptraj_init.getLength()
        assert (traj_size >= 2)
        start_dmpstate = critical_dmptraj_init.getDMPStateAtIndex(0)
        goal_dmpstate = critical_dmptraj_init.getDMPStateAtIndex(traj_size-1)
        if (tau_init == None):
            if (robot_task_servo_rate == None):
                self.tau = goal_dmpstate.getTime()[0,0] - start_dmpstate.getTime()[0,0]
            else:
                self.tau = (1.0 * (traj_size-1))/robot_task_servo_rate
        else:
            self.tau = tau_init
        assert (self.tau >= MIN_TAU), "DMPUnrollInitParams.tau=" + str(self.tau) + " < MIN_TAU"
        critical_dmpstates_list = []
        critical_dmpstates_list.append(start_dmpstate)
        if (traj_size >= 3):
            before_goal_dmpstate = critical_dmptraj_init.getDMPStateAtIndex(traj_size-2)
            critical_dmpstates_list.append(before_goal_dmpstate)
        critical_dmpstates_list.append(goal_dmpstate)
        processed_critical_states_list = []
        for critical_dmpstate in critical_dmpstates_list:
            if (is_zeroing_out_velocity_and_acceleration):
                processed_critical_states_list.append(DMPState(critical_dmpstate.getX()))
            else:
                processed_critical_states_list.append(copy.copy(critical_dmpstate))
        self.critical_states = convertDMPStatesListIntoDMPTrajectory(processed_critical_states_list)
    
    def isValid(self):
        assert (self.tau >= MIN_TAU), "DMPUnrollInitParams.tau=" + str(self.tau) + " < MIN_TAU"
        assert (self.critical_states.getLength() >= 2)
        assert (self.critical_states.isValid())
        return True

def getDMPUnrollInitParams(start_dmpstate, goal_dmpstate, tau_init=None, robot_task_servo_rate=None, is_zeroing_out_velocity_and_acceleration=True):
    critical_dmptraj_init = convertDMPStatesListIntoDMPTrajectory([start_dmpstate, goal_dmpstate])
    return DMPUnrollInitParams(critical_dmptraj_init, tau_init, robot_task_servo_rate, 
                               is_zeroing_out_velocity_and_acceleration)

File: src/dmp_param/DMPUnrollInitParams.cpp
#include "dmp/dmp_param/DMPUnrollInitParams.h"

namespace dmp {

DMPUnrollInitParams::DMPUnrollInitParams() : tau(0.0), rt_assertor(NULL) {}

DMPUnrollInitParams::DMPUnrollInitParams(
    const double& tau_init, const DMPState& start_state,
    const DMPState& goal_state, RealTimeAssertor* real_time_assertor,
    bool is_zeroing_out_velocity_and_acceleration)
    : tau(tau_init), rt_assertor(real_time_assertor) {
  critical_states.resize(2);
  if (is_zeroing_out_velocity_and_acceleration) {
    critical_states[0] = DMPState(start_state.getX(), rt_assertor);
    critical_states[1] = DMPState(goal_state.getX(), rt_assertor);
  } else {
    critical_states[0] = start_state;
    critical_states[1] = goal_state;
  }
}

DMPUnrollInitParams::DMPUnrollInitParams(
    const double& tau_init, const Trajectory& trajectory,
    RealTimeAssertor* real_time_assertor,
    bool is_zeroing_out_velocity_and_acceleration)
    : tau(tau_init), rt_assertor(real_time_assertor) {
  uint traj_size = trajectory.size();

  if (traj_size >= 2) {
    if (traj_size >= 3) {
      critical_states.resize(3);
    } else {
      critical_states.resize(2);
    }

    if (critical_states.size() >= 1) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[0] = DMPState(trajectory[0].getX(), rt_assertor);
      } else {
        critical_states[0] = trajectory[0];
      }
    }

    if (critical_states.size() >= 2) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 1] =
            DMPState(trajectory[traj_size - 1].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 1] = trajectory[traj_size - 1];
      }
    }

    if (critical_states.size() >= 3) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 2] =
            DMPState(trajectory[traj_size - 2].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 2] = trajectory[traj_size - 2];
      }
    }
  }
}

DMPUnrollInitParams::DMPUnrollInitParams(
    const Trajectory& trajectory, RealTimeAssertor* real_time_assertor,
    bool is_zeroing_out_velocity_and_acceleration)
    : rt_assertor(real_time_assertor) {
  uint traj_size = trajectory.size();

  if (traj_size >= 2) {
    // if the trajectory is containing the "correct" information of the
    // trajectory timing, then use this timing information to compute tau:
    tau = trajectory[traj_size - 1].getTime() - trajectory[0].getTime();

    if (traj_size >= 3) {
      critical_states.resize(3);
    } else {
      critical_states.resize(2);
    }

    if (critical_states.size() >= 1) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[0] = DMPState(trajectory[0].getX(), rt_assertor);
      } else {
        critical_states[0] = trajectory[0];
      }
    }

    if (critical_states.size() >= 2) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 1] =
            DMPState(trajectory[traj_size - 1].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 1] = trajectory[traj_size - 1];
      }
    }

    if (critical_states.size() >= 3) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 2] =
            DMPState(trajectory[traj_size - 2].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 2] = trajectory[traj_size - 2];
      }
    }
  }
}

DMPUnrollInitParams::DMPUnrollInitParams(
    const Trajectory& trajectory, const double& robot_task_servo_rate,
    RealTimeAssertor* real_time_assertor,
    bool is_zeroing_out_velocity_and_acceleration)
    : rt_assertor(real_time_assertor) {
  uint traj_size = trajectory.size();

  if (traj_size >= 2) {
    // if the trajectory is containing the "correct" information of the
    // trajectory timing, then use this timing information to compute tau:
    tau = trajectory[traj_size - 1].getTime() - trajectory[0].getTime();
    // otherwise use the robot_task_servo_rate to compute tau
    // (assuming that the trajectory is sampled at robot_task_servo_rate):
    if (tau < MIN_TAU) {
      tau = (1.0 * (traj_size - 1)) / robot_task_servo_rate;
    }

    if (traj_size >= 3) {
      critical_states.resize(3);
    } else {
      critical_states.resize(2);
    }

    if (critical_states.size() >= 1) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[0] = DMPState(trajectory[0].getX(), rt_assertor);
      } else {
        critical_states[0] = trajectory[0];
      }
    }

    if (critical_states.size() >= 2) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 1] =
            DMPState(trajectory[traj_size - 1].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 1] = trajectory[traj_size - 1];
      }
    }

    if (critical_states.size() >= 3) {
      if (is_zeroing_out_velocity_and_acceleration) {
        critical_states[critical_states.size() - 2] =
            DMPState(trajectory[traj_size - 2].getX(), rt_assertor);
      } else {
        critical_states[critical_states.size() - 2] = trajectory[traj_size - 2];
      }
    }
  }
}

bool DMPUnrollInitParams::isValid() const {
  if (rt_assert(tau >= MIN_TAU) == false) {
    return false;
  }
  if (rt_assert(critical_states.size() >= 2) == false) {
    return false;
  }
  for (uint i = 0; i < critical_states.size(); i++) {
    if (rt_assert(critical_states[i].isValid()) == false) {
      return false;
    }
  }
  return true;
}

DMPUnrollInitParams::~DMPUnrollInitParams() {}

}  // namespace dmp
